### PR TITLE
Allow collection expressions with `CreateCompilation`

### PR DESCRIPTION
**Pull request by jcouv**

### CODE REVIEW

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler Small PR for review. Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler Small PR for review. Thanks

File: src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs [hunk 1]
```diff
@@ -83,5 +84,17 @@ public SyntaxTree[] GetSyntaxTrees(CSharpParseOptions parseOptions, string sourc
         public static implicit operator CSharpTestSource(List<SyntaxTree> source) => new CSharpTestSource(source.ToArray());
         public static implicit operator CSharpTestSource(ImmutableArray<SyntaxTree> source) => new CSharpTestSource(source.ToArray());
         public static implicit operator CSharpTestSource(CSharpTestSource[] source) => new CSharpTestSource(source);
+
+        // Dummy IEnumerable support to satisfy the collection expression and CollectionBuilder requirements
+        IEnumerator<CSharpTestSource> IEnumerable<CSharpTestSource>.GetEnumerator() => throw new NotImplementedException();
+        IEnumerator IEnumerable.GetEnumerator() => throw new NotImplementedException();
+
+        internal class CSharpTestSourceBuilder
```
Comment (jjonescz) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: Can this be static? Or sealed at least.

**Review by jjonescz — APPROVED** [review]

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/Test/Core/CollectionBuilderAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if !NET8_0_OR_GREATER
```
Comment (333fred) [inline src/Compilers/Test/Core/CollectionBuilderAttribute.cs:5]: Just to make sure that we don't run afoul of any weird situations (such as razor test dependencies), we should typeforward in an `#else`: https://github.com/dotnet/roslyn/blob/main/docs/Target%20Framework%20Strategy.md#pattern-for-types

File: src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs [hunk 3]
```diff
@@ -83,5 +84,17 @@ public SyntaxTree[] GetSyntaxTrees(CSharpParseOptions parseOptions, string sourc
         public static implicit operator CSharpTestSource(List<SyntaxTree> source) => new CSharpTestSource(source.ToArray());
         public static implicit operator CSharpTestSource(ImmutableArray<SyntaxTree> source) => new CSharpTestSource(source.ToArray());
         public static implicit operator CSharpTestSource(CSharpTestSource[] source) => new CSharpTestSource(source);
+
+        // Dummy IEnumerable support to satisfy the collection expression and CollectionBuilder requirements
+        IEnumerator<CSharpTestSource> IEnumerable<CSharpTestSource>.GetEnumerator() => throw new NotImplementedException();
+        IEnumerator IEnumerable.GetEnumerator() => throw new NotImplementedException();
+
+        internal static class CSharpTestSourceBuilder
```
Comment (333fred) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: Why does this need to be a separate type? Why not just make it a static member on this type?

Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: It does not *need* to be a separate type, but it is customary for the collection and the collection builder to be different types, and I think it makes sense. See all the collection tests

File: src/Compilers/Test/Core/CollectionBuilderAttribute.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/Test/Core/CollectionBuilderAttribute.cs:5]: Thanks for the pointer. That's a useful doc

File: src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: > but it is customary for the collection and the collection builder to be different types

This is a brand new pattern, there's no "customary" at all 😆. I'd just as soon keep it simpler.

Comment (jcouv) [reply src/Compilers/Test/Utilities/CSharp/CSharpTestSource.cs:92]: All our collection builder tests beg to differ. The spec also calls out [two different concepts](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#create-methods): a "collection type" and a "builder type". It is clearer not to blend them and doesn't seem "simpler".
In any case, this seems a nitpick/preference, not a reason to hold this change up. Thanks

**Review by jaredpar — APPROVED** [review]

**Pull request merged**